### PR TITLE
feat: add formdata and octet stream support

### LINKED ISSUE #5659: bug: request body read multiple times
### Provide environment information

- Latest canary
- `11.0.0-rc.334` in fine if you encountered this issue

### Describe the bug

Some error akin to

> error: Body already used

### Link to reproduction

https://discord.com/channels/867764511159091230/1231690793799127141/1231690793799127141

### To reproduce

n/a

### Additional information

creating this to track regression tests

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5659">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5659/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Nick-Lucas**
Closes #5659

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/server/src/adapters/fetch/content-type/json/index.ts [hunk 1]
```diff
@@ -16,79 +17,75 @@ export interface FetchHTTPContentTypeHandler<TRouter extends AnyRouter>
...
-          ?.startsWith('application/json')
-          ? await opts.req.text()
-          : '';
+            return JSON.parse(input);
+          }
 
-        if (typeof body === 'string') {
-          // A mutation with no inputs will have req.body === ''
-          return body.length === 0 ? undefined : JSON.parse(body);
+          return await opts.req.json();
```
Comment (Nick-Lucas) [inline packages/server/src/adapters/fetch/content-type/json/index.ts:41]: "Use the platform" - @juliusmarminge

File: packages/server/src/adapters/fetch/content-type/json/index.ts [hunk 2]
```diff
@@ -16,79 +17,75 @@ export interface FetchHTTPContentTypeHandler<TRouter extends AnyRouter>
...
-        }
+  return {
+    name: 'fetch-json',
+    isMatch: (headers) => {
+      return !!headers.get('content-type')?.startsWith('application/json');
+    },
+    getInputs: async (opts, info) => {
+      async function getRawProcedureInputOrThrow() {
+        try {
+          if (opts.url.searchParams.has('input')) {
```
Comment (Nick-Lucas) [inline packages/server/src/adapters/fetch/content-type/json/index.ts:32]: Probably better with possible method overrides?

File: packages/server/src/adapters/fetch/content-type/json/index.ts [hunk 3]
```diff
@@ -16,79 +17,75 @@ export interface FetchHTTPContentTypeHandler<TRouter extends AnyRouter>
...
->() => FetchHTTPContentTypeHandler<TRouter> = () => ({
-  name: 'fetch-json',
-  isMatch: (headers) => {
-    return !!headers.get('content-type')?.startsWith('application/json');
-  },
-  getInputs: async (opts, info) => {
-    async function getRawProcedureInputOrThrow() {
-      const { req } = opts;
+>() => FetchHTTPContentTypeHandler<TRouter> = () => {
+  const rawInputCache = createConcurrentCache();
```
Comment (Nick-Lucas) [inline packages/server/src/adapters/fetch/content-type/json/index.ts:21]: The bug was that when we map over a cache, all the promises which start waiting on this body start at once and pass control to the next one while waiting on the body. So yes we were consuming this N times at once.

By using a concurrent cache the first iteration will synchronously cache the promise and then future batches will read the current promise, additionally we can move the batch caching inside here (see next line) which is better for simplicity in the adapter handlers

File: packages/tests/server/adapters/fetch.test.ts [hunk 4]
```diff
@@ -266,3 +269,22 @@ test.each([
...
+    expect(res).toEqual(['hello world', 'hello KATT']);
+  } catch (err) {
+    console.log('err', err);
+    console.log('err', (err as any).cause);
+
+    throw err;
+  } finally {
+    await t.close();
+  }
+});
```
Comment (Nick-Lucas) [inline packages/tests/server/adapters/fetch.test.ts:290]: No idea why my repro attempt last night didn't trigger the bug but this does? Would love to understand

Comment (KATT) [reply packages/tests/server/adapters/fetch.test.ts:290]: I'm not sure but in the snippet you shared you used a query and not a mutation, a query doesn't read the body

File: packages/server/src/adapters/fetch/content-type/json/index.ts [hunk 5]
```diff
@@ -16,79 +17,74 @@ export interface FetchHTTPContentTypeHandler<TRouter extends AnyRouter>
...
->() => FetchHTTPContentTypeHandler<TRouter> = () => ({
-  name: 'fetch-json',
-  isMatch: (headers) => {
-    return !!headers.get('content-type')?.startsWith('application/json');
-  },
-  getInputs: async (opts, info) => {
-    async function getRawProcedureInputOrThrow() {
-      const { req } = opts;
+>() => FetchHTTPContentTypeHandler<TRouter> = () => {
+  const cache = createConcurrentCache();
```
Comment (KATT) [inline packages/server/src/adapters/fetch/content-type/json/index.ts:21]: Couldn't this just be 

```suggestion
  const getRawProcedureInputOrThrow = memoize(async () => {
    // ...impl
  })
```

it's more straight-forward than the concurrent cache function where the key value etc

it's easy to write a memo-fn

```ts

function memoize<TReturn>(fn: () => TReturn): () => TReturn {
  let cache: TReturn | undefined;
  return () => {
    if (!cache) {
      cache = fn();
    }
    return cache;
  };
}
```

Comment (Nick-Lucas) [reply packages/server/src/adapters/fetch/content-type/json/index.ts:21]: How would that work for individual batch numbers?

Comment (KATT) [reply packages/server/src/adapters/fetch/content-type/json/index.ts:21]: We don't need to memoize the deserialization, do we? I

Comment (Nick-Lucas) [reply packages/server/src/adapters/fetch/content-type/json/index.ts:21]: So it won’t explode, but it is unnecessary work and getRawInput might be called multiple times depending on setup. Particularly for larger batches it’s a low hanging optimisation (I really just moved it inside here from resolveHTTPResponse to simplify that method)

I like the memo syntax you’re suggesting btw, and we could probably do it all in a more raw way

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Review by KATT — DISMISSED** [review]

**Review by KATT — APPROVED** [review]

**Pull request merged**